keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,415 @@
+   susceptible to this risk.
+
+#### Mitigations
+
+1. Such problems are expected to surface during the Beta phase when the feature
+   gate will be enabled by default. The mitigation at that point would be to set
+   the feature gate to "off" and default back to current behavior. We will also
+   make the graduation criteria _to_ Beta be the document we would like to write
+   and mention this as an explicit recommendation of what not to do when
+   deploying Kube-proxy. As such any vendor doing this would get a heads-up
+   during Alpha.
+
+2. If Kube-proxy starts failing when reading from the API server, it should just
+   assume that the last state seen continues. For Kube-proxy to be made aware of
+   this, it needs to invoke
+   `serviceInformer.Informer().SetWatchErrorHandler(DefaultWatchErrorHandler)`
+   when initializing its informers. Any errors observed by client-go when
+   watching from the API server will be reported on `DefaultWatchErrorHandler`.
+
+3. Metrics should inform on Kube-proxy health and include information about its
+   `healthz` state, this can then be used to correlate to networking metrics
+   surrounding new/established connections on the node. Ex: a failing `healthz`
+   should correlate to a total drop in the count of new connections and with a
+   zero-or-negative rate of established connections. E2E tests should also be
+   designed with this specific goal in mind, i.e: validating the impact of a
+   failing kube-proxy on ingress connectivity. Kube-proxy currently has a lot of
+   metrics regarding how its health is doing, but no direct red/green indicator
+   of what the end result of its health is. One such metric could be
+   `proxy_health_check_200_count` / `proxy_health_check_503_count`
+
+4. The feature could be disabled for user who is dependent upon such behavior by
+   means of flipping the feature flag to off.
+
+## Design Details
+
+1. Implement Kube-proxy change invoking client-go's `SetWatchErrorHandler`  on
+   watch errors from the API server. This addresses the second point in
+   [Mitigations](#mitigations)
+
+2. Implement change in Kube-proxy which will react to changes on the Node object
+   and once the taint `ToBeDeletedByClusterAutoscaler` is placed on the Node
+   object: start failing it's `healthz` state.
+
+3. Write document to be published at: https://kubernetes.io/docs/concepts/ which
+   details: a) how determining node/instance health can best be done for
+   Kubernetes clusters b) how Kube-proxy will do it once the changes proposed in
+   this KEP are merged c) what some pitfalls with other methods might be.
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+Update the Kube-proxy unit tests to include the `healthz` answer for the
+healthcheck server test suite.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
```

Comment:
While arguably true, we're really interested here in how the changes specific to this KEP will be covered, and we can do better than the existing code 🙃 

I think a more relevant response is more like "this feature is not readily integration tested, so we will use unit and e2e"